### PR TITLE
[Doc] Add docstring for init(), reset() and a few constants

### PR DESCRIPTION
**Pull request by k-ye**
Related issue = #

### CODE REVIEW

File: python/taichi/lang/__init__.py [hunk 1]
```diff
@@ -61,16 +61,42 @@
 
 cfg = default_cfg()
 x86_64 = _ti_core.x64
+"""The X64 CPU backend.
+"""
 x64 = _ti_core.x64
```
Comment (qiao-bo) [inline python/taichi/lang/__init__.py:66]: A short question: why do we have two names for _ti_core.x64 ?

File: python/taichi/lang/__init__.py [hunk 2]
```diff
@@ -286,7 +316,24 @@ def init(arch=None,
...
          **kwargs):
+    """Initializes the Taichi runtime.
+
+    This should always be the entry point of your Taichi program. Most
+    importantly, it sets the backend used throughout the program.
 
+    Args:
+        arch: Backend to use. This is usually :const:`~taichi.lang.cpu` or :const:`~taichi.lang.gpu`.
+        default_fp (Optional[type]): Default floating-point type.
+        default_fp (Optional[type]): Default integral type.
```
Comment (qiao-bo) [inline python/taichi/lang/__init__.py:327]: default_ip?

Comment (qiao-bo) [inline python/taichi/lang/__init__.py:327]:
```suggestion
        default_ip (Optional[type]): Default integral type.
```

File: python/taichi/lang/__init__.py [hunk 3]
```diff
@@ -286,7 +316,24 @@ def init(arch=None,
...
+        **kwargs: Taichi provides a large amount of flags for configuration. Below we list a few
```
Comment (qiao-bo) [inline python/taichi/lang/__init__.py:328]: a large amount of -> a large number of

File: python/taichi/lang/__init__.py [hunk 4]
```diff
@@ -286,7 +316,24 @@ def init(arch=None,
...
+            most frequently used ones. For a complete list, please check out
+            https://github.com/taichi-dev/taichi/blob/master/taichi/program/compile_config.h.
+
+            * ``cpu_max_num_threads`` (int): Sets the number of threads used by the CPU thread pool.
+            * ``debug`` (bool): Enable the debug mode, under which Taichi does a few more things like boundary checks.
```
Comment (qiao-bo) [inline python/taichi/lang/__init__.py:333]:
```suggestion
            * ``debug`` (bool): Enables the debug mode, under which Taichi does a few more things like boundary checks.
```

File: python/taichi/lang/__init__.py (hunk 3 above)
Comment (qiao-bo) [inline python/taichi/lang/__init__.py:328]:
```suggestion
        **kwargs: Taichi provides a large number of flags for configuration. Below we list a few
```

File: python/taichi/lang/__init__.py [hunk 5]
```diff
@@ -286,7 +316,24 @@ def init(arch=None,
...
+            * ``print_ir`` (bool): Print the CHI IR of the Taichi kernels
```
Comment (qiao-bo) [inline python/taichi/lang/__init__.py:334]:
```suggestion
            * ``print_ir`` (bool): Prints the CHI IR of the Taichi kernels
```

**Review by qiao-bo — APPROVED** [review]
Again a few nits from my side ;)

File: python/taichi/lang/__init__.py (hunk 1 above)
Comment (k-ye) [reply python/taichi/lang/__init__.py:66]: likely a technical debt...

File: python/taichi/lang/__init__.py [hunk 6]
```diff
@@ -286,7 +316,24 @@ def init(arch=None,
...
+    """Initializes the Taichi runtime.
+
+    This should always be the entry point of your Taichi program. Most
+    importantly, it sets the backend used throughout the program.
 
+    Args:
+        arch: Backend to use. This is usually :const:`~taichi.lang.cpu` or :const:`~taichi.lang.gpu`.
+        default_fp (Optional[type]): Default floating-point type.
+        default_ip (Optional[type]): Default integral type.
+        **kwargs: Taichi provides a large number of flags for configuration. Below we list a few
```
Comment (ailzhang) [inline python/taichi/lang/__init__.py:328]: maybe say this in another way? ;) Taichi provides highly customizable compilation through kwargs which allows fine grained control of Taichi compiler behavior.

File: python/taichi/lang/__init__.py [hunk 7]
```diff
@@ -398,6 +445,11 @@ def no_activate(*args):
 
 
 def block_local(*args):
+    """Hints Taichi to cache the list of fields into the shared memory.
```
Comment (ailzhang) [inline python/taichi/lang/__init__.py:448]: Do we also want to mention that it can potentially bring some speedup but it's not guaranteed?

**Review by ailzhang — APPROVED** [review]
Looks great! Thanks!

File: python/taichi/lang/__init__.py (hunk 6 above)
Comment (k-ye) [reply python/taichi/lang/__init__.py:328]: SG!

File: python/taichi/lang/__init__.py (hunk 7 above)
Comment (k-ye) [reply python/taichi/lang/__init__.py:448]: Added the link to the doc explaining BLS

File: python/taichi/lang/__init__.py [hunk 8]
```diff
@@ -61,16 +61,42 @@
 
 cfg = default_cfg()
 x86_64 = _ti_core.x64
+"""The X64 CPU backend.
```
Comment (strongoier) [inline python/taichi/lang/__init__.py:64]:
```suggestion
"""The x64 CPU backend.
```

File: python/taichi/lang/__init__.py [hunk 9]
```diff
@@ -286,7 +316,26 @@ def init(arch=None,
...
+    This should always be the entry point of your Taichi program. Most
+    importantly, it sets the backend used throughout the program.
 
+    Args:
+        arch: Backend to use. This is usually :const:`~taichi.lang.cpu` or :const:`~taichi.lang.gpu`.
+        default_fp (Optional[type]): Default floating-point type.
+        default_fp (Optional[type]): Default integral type.
+        **kwargs: Taichi provides highly customizable compilation through
+            ``kwargs``, which allows for fine grained control of Taichi compiler
+            behavior. Below we list some of the most frequently used ones.For a
```
Comment (strongoier) [inline python/taichi/lang/__init__.py:330]:
```suggestion
            behavior. Below we list some of the most frequently used ones. For a
```

File: python/taichi/lang/__init__.py [hunk 10]
```diff
@@ -286,7 +316,26 @@ def init(arch=None,
...
+            complete list, please check out
+            https://github.com/taichi-dev/taichi/blob/master/taichi/program/compile_config.h.
+
+            * ``cpu_max_num_threads`` (int): Sets the number of threads used by the CPU thread pool.
+            * ``debug`` (bool): Enables the debug mode, under which Taichi does a few more things like boundary checks.
+            * ``print_ir`` (bool): Prints the CHI IR of the Taichi kernels
```
Comment (strongoier) [inline python/taichi/lang/__init__.py:336]:
```suggestion
            * ``print_ir`` (bool): Prints the CHI IR of the Taichi kernels.
```

File: python/taichi/lang/__init__.py [hunk 11]
```diff
@@ -286,7 +316,26 @@ def init(arch=None,
...
+            * ``packed`` (bool): Enables the packed memory layout. See https://docs.taichi.graphics/docs/lang/articles/advanced/layout
```
Comment (strongoier) [inline python/taichi/lang/__init__.py:337]:
```suggestion
            * ``packed`` (bool): Enables the packed memory layout. See https://docs.taichi.graphics/docs/lang/articles/advanced/layout.
```

File: python/taichi/lang/__init__.py [hunk 12]
```diff
@@ -398,6 +447,18 @@ def no_activate(*args):
 
 
 def block_local(*args):
+    """Hints Taichi to cache the fields and to enable the BLS optimization.
+
+    Please visit https://docs.taichi.graphics/docs/lang/articles/advanced/performance
+    for how BLS is used.
+
+    Args:
+        *args (List[Taichi field]): A list of sparse Taichi fields.
```
Comment (strongoier) [inline python/taichi/lang/__init__.py:456]:
```suggestion
        *args (List[Field]): A list of sparse Taichi fields.
```

**Review by strongoier — APPROVED** [review]
LGTM! Only a few nits.

**Pull request merged**